### ISSUE TITLE
runtime/fetcher needs auxiliary function for retriable pipeline downloads

### ISSUE THREAD

**Issue by jonasfj** [taskcluster/taskcluster-worker#366]
Sometimes we want to use the `Fetcher` interface, but process the downloaded data in a pipeline, instead of a file. Obviously, such a pipeline has to be resetable...

```golang
func processIncoming(ctx context.Context, stream io.Reader) error

var target fetcher.WriteReseter = fetcher.Pipeline(processIncoming)
```

In `fetcher` we would need:
```golang
func Pipeline(makePipeline func(ctx context.Context, stream io.Reader) error) WriteReseter {
  ...
}
```

This could be done as follows: https://gist.github.com/jonasfj/74144fc09d25a4aa2a0da5d58f152eb1

**Use cases**:
 * download a docker image and do the zstd decompression, image rename and docker load on the fly.
 * same thing for qemu-engine, which currently uses an intermediate file..

**Comment by jonasfj** [comment]
Note. We probably still need a way to wait for the pipeline to be completed after fetcher returns.

**Comment by jonasfj** [comment]
This landed in #369
